File: Game-1-modular/entities/components/equipment_manager.py
"""Equipment management component"""

from typing import Dict, Optional, Tuple

from data.models import EquipmentItem


class EquipmentManager:
    def __init__(self):
        self.slots = {
            'mainHand': None,
            'offHand': None,
            'helmet': None,
            'chestplate': None,
            'leggings': None,
            'boots': None,
            'gauntlets': None,
            'accessory': None,
            'axe': None,  # Tool slot for axes
            'pickaxe': None,  # Tool slot for pickaxes
        }

    def equip(self, item: EquipmentItem, character) -> Tuple[Optional[EquipmentItem], str]:
        """Equip an item, returns (previously_equipped_item, status_message)"""
        print(f"      🔧 EquipmentManager.equip() called")
        print(f"         - item: {item.name} (slot: {item.slot}, hand_type: {item.hand_type})")

        can_equip, reason = item.can_equip(character)
        print(f"         - can_equip: {can_equip}, reason: {reason}")

        if not can_equip:
            print(f"         ❌ Cannot equip: {reason}")
            return None, reason

        slot = item.slot
        if slot not in self.slots:
            print(f"         ❌ Invalid slot '{slot}' not in {list(self.slots.keys())}")
            return None, f"Invalid slot: {slot}"

        # HAND TYPE VALIDATION
        if slot == 'mainHand':
            # Note: 2H weapon offhand unequip is handled by try_equip_from_inventory
            # This validation is just a safety check
            if item.hand_type == "2H":
                if self.slots['offHand'] is not None:
                    print(f"         ⚠️ Warning: 2H weapon equipped with offhand still occupied")
                    print(f"            This should have been auto-unequipped by caller")
        elif slot == 'offHand':
            # Check if mainhand allows offhand
            mainhand = self.slots['mainHand']
            if mainhand is not None:
                if mainhand.hand_type == "2H":
                    return None, "Cannot equip offhand - mainhand is 2H weapon"
                elif mainhand.hand_type == "default":
                    # Default weapons can't have offhand unless this is a shield
                    if item.item_type != "shield":
                        return None, "Mainhand weapon doesn't support offhand"
                elif mainhand.hand_type == "versatile":
                    # Versatile weapons can have 1H in offhand, but not versatile/2H
                    if item.item_type != "shield" and item.hand_type != "1H":
                        return None, "Versatile mainhand only allows 1H or shield in offhand"

            # Offhand can only equip 1H items or shields (NOT versatile or 2H)
            if item.hand_type not in ["1H"] and item.item_type != "shield":
                return None, "Item cannot be equipped in offhand (must be 1H or shield)"

        old_item = self.slots[slot]
        self.slots[slot] = item
        print(f"         ✅ Equipped to slot '{slot}'")

        # Debug: Show weapon damage if equipping to weapon slot
        if slot in ['mainHand', 'offHand']:
            print(f"         🗡️  Weapon damage: {item.damage}")
            print(f"         🗡️  Actual damage (with effectiveness): {item.get_actual_damage()}")

        # Recalculate character stats
        character.recalculate_stats()

        # Debug: Show new weapon damage total
        if slot in ['mainHand', 'offHand']:
            weapon_dmg = self.get_weapon_damage()
            print(f"         🎯 Total weapon damage range: {weapon_dmg}")

        return old_item, "OK"

    def unequip(self, slot: str, character) -> Optional[EquipmentItem]:
        """Unequip item from slot"""
        if slot not in self.slots:
            return None
        item = self.slots[slot]
        self.slots[slot] = None

        # Recalculate character stats
        character.recalculate_stats()

        return item

    def is_equipped(self, item_id: str) -> bool:
        """Check if an item is currently equipped"""
        for item in self.slots.values():
            if item and item.item_id == item_id:
                return True
        return False

    def get_total_defense(self) -> int:
        """Get total defense from all armor pieces including enchantment effects"""
        total = 0
        armor_slots = ['helmet', 'chestplate', 'leggings', 'boots', 'gauntlets']
        for slot in armor_slots:
            item = self.slots.get(slot)
            if item:
                total += item.get_defense_with_enchantments()
        return total

    def get_weapon_damage(self, hand: str = 'mainHand') -> Tuple[int, int]:
        """Get damage from specified hand (mainHand or offHand)"""
        weapon = self.slots.get(hand)
        if weapon:
            return weapon.get_actual_damage()
        if hand == 'mainHand':
            return (1, 2)  # Unarmed damage for mainhand
        return (0, 0)  # No offhand

    def get_weapon_range(self, hand: str = 'mainHand') -> float:
        """Get range of equipped weapon in specified hand, default to 1.0 for unarmed"""
        weapon = self.slots.get(hand)
        if weapon:
            base_range = weapon.range

            # Add tag-based range bonus (reach)
            weapon_tags = weapon.get_metadata_tags()
            if weapon_tags:
                from entities.components.weapon_tag_calculator import WeaponTagModifiers
                range_bonus = WeaponTagModifiers.get_range_bonus(weapon_tags)
                return base_range + range_bonus

            return base_range
        if hand == 'mainHand':
            return 1.0  # Unarmed/default range
        return 0.0  # No offhand range

    def get_weapon_attack_speed(self, hand: str = 'mainHand') -> float:
        """Get attack speed multiplier of weapon in specified hand"""
        weapon = self.slots.get(hand)
        if weapon:
            return weapon.attack_speed
        return 1.0  # Default attack speed

    def get_stat_bonuses(self) -> Dict[str, float]:
        bonuses = {}
        for item in self.slots.values():
            if item:
                for stat, value in item.bonuses.items():
                    bonuses[stat] = bonuses.get(stat, 0) + value
        return bonuses

File: Game-1-Godot/src/Game1.Core/Data/EquipmentManager.cs
using Game1.Core.Combat;
using Game1.Core.Progression;

namespace Game1.Core.Data;

/// <summary>
/// Port of entities/components/equipment_manager.py — 8 equipment slots + 2
/// tool slots, hand-type validation matrix (2H blocks offhand; default allows
/// shield only; versatile allows 1H or shield), defense/damage/range/attack-
/// speed/stat-bonus aggregation. Stat recalculation and the EQUIPMENT_CHANGED
/// bus publish become the Changed event.
/// </summary>
public sealed class EquipmentManager
{
    private static readonly string[] SlotNames =
    {
        "mainHand", "offHand", "helmet", "chestplate", "leggings",
        "boots", "gauntlets", "accessory", "axe", "pickaxe",
    };

    public Dictionary<string, EquipmentItem?> Slots { get; } =
        SlotNames.ToDictionary(s => s, _ => (EquipmentItem?)null);

    /// <summary>(itemId, slot, equipped) — replaces stat_tracker + bus publish.</summary>
    public event Action<string, string, bool>? Changed;

    // equipment_manager.py:23-97
    public (EquipmentItem? OldItem, string Reason) Equip(
        EquipmentItem item, ICharacterQuery character)
    {
        var (canEquip, reason) = item.CanEquip(character);
        if (!canEquip)
            return (null, reason);

        var slot = item.Slot;
        if (!Slots.ContainsKey(slot))
            return (null, $"Invalid slot: {slot}");

        if (slot == "offHand")
        {
            var mainhand = Slots["mainHand"];
            if (mainhand is not null)
            {
                if (mainhand.HandType == "2H")
                    return (null, "Cannot equip offhand - mainhand is 2H weapon");
                if (mainhand.HandType == "default" && item.ItemType != "shield")
                    return (null, "Mainhand weapon doesn't support offhand");
                if (mainhand.HandType == "versatile"
                    && item.ItemType != "shield" && item.HandType != "1H")
                    return (null, "Versatile mainhand only allows 1H or shield in offhand");
            }
            if (item.HandType != "1H" && item.ItemType != "shield")
                return (null, "Item cannot be equipped in offhand (must be 1H or shield)");
        }

        var oldItem = Slots[slot];
        Slots[slot] = item;
        Changed?.Invoke(item.ItemId, slot, true);
        return (oldItem, "OK");
    }

    // :99-123
    public EquipmentItem? Unequip(string slot)
    {
        if (!Slots.TryGetValue(slot, out var item))
            return null;
        Slots[slot] = null;
        if (item is not null)
            Changed?.Invoke(item.ItemId, slot, false);
        return item;
    }

    public bool IsEquipped(string itemId) =>
        Slots.Values.Any(i => i is not null && i.ItemId == itemId);

    // :132-140
    public int GetTotalDefense() =>
        new[] { "helmet", "chestplate", "leggings", "boots", "gauntlets" }
            .Select(s => Slots[s])
            .Where(i => i is not null)
            .Sum(i => i!.GetDefenseWithEnchantments());

    // :142-149 — unarmed mainhand (1,2); empty offhand (0,0)
    public (int Min, int Max) GetWeaponDamage(string hand = "mainHand")
    {
        var weapon = Slots.GetValueOrDefault(hand);
        if (weapon is not null)
            return weapon.GetActualDamage();
        return hand == "mainHand" ? (1, 2) : (0, 0);
    }

    // :151-171 — any held item +1 range over fists; reach tag adds on top
    public double GetWeaponRange(string hand = "mainHand")
    {
        var weapon = Slots.GetValueOrDefault(hand);
        if (weapon is not null)
        {
            var baseRange = weapon.Range + 1.0;
            var tags = weapon.Tags;
            if (tags.Count > 0)
                return baseRange + WeaponTagModifiers.GetRangeBonus(tags);
            return baseRange;
        }
        return hand == "mainHand" ? 1.0 : 0.0;
    }

    public double GetWeaponAttackSpeed(string hand = "mainHand") =>
        Slots.GetValueOrDefault(hand)?.AttackSpeed ?? 1.0;

    // :180-186 — additive across all equipped items
    public Dictionary<string, double> GetStatBonuses()
    {
        var bonuses = new Dictionary<string, double>();
        foreach (var item in Slots.Values)
        {
            if (item is null) continue;
            foreach (var kv in item.Bonuses)
            {
                var value = J.AsNum(kv.Value) ?? 0;
                bonuses[kv.Key] = bonuses.GetValueOrDefault(kv.Key, 0) + value;
            }
        }
        return bonuses;
    }
}
